File: neo4j_util_go/neo4j_utils_main.go
package neo4j_util_go

import (
	"fmt"
	"log"
	"time"
)

// CreateGovernmentNode creates a government node with retry mechanism
func CreateGovernmentNode(neo4j *Neo4jInterface) ([]map[string]interface{}, error) {
	query := `
	MERGE (g:government {id: $id})
	SET g.name = $name
	RETURN g
	`
	params := map[string]interface{}{
		"id":   "gov_01",
		"name": "Government of Sri Lanka",
	}

	maxAttempts := 3
	var lastErr error

	for attempt := 1; attempt <= maxAttempts; attempt++ {
		fmt.Printf("Attempt %d of %d to create government node...\n", attempt, maxAttempts)
		result, err := neo4j.ExecuteQuery(query, params)
		if err == nil {
			fmt.Println("Successfully created government node.")
			return result, nil
		}

		lastErr = err
		fmt.Printf("Error on attempt %d: %v\n", attempt, err)

		if attempt < maxAttempts {
			waitTime := time.Duration(2*attempt) * time.Second
			fmt.Printf("Waiting %v seconds before retrying...\n", waitTime.Seconds())
			time.Sleep(waitTime)
		}
	}

	return nil, fmt.Errorf("maximum retry attempts reached. Failed to create government node: %w", lastErr)
}

// RunNeo4jUtil is the main entry point for the Neo4j utility
func RunNeo4jUtil() {
	// Initialize Neo4j interface
	neo4jInterface, err := NewNeo4jInterface()
	if err != nil {
		log.Fatalf("Failed to initialize Neo4j interface: %v", err)
	}

	// Ensure we always close the connection
	defer neo4jInterface.Close()

	// Create the government node
	govNode, err := CreateGovernmentNode(neo4jInterface)
	if err != nil {
		log.Printf("Error creating government node: %v", err)
		return
	}

	// Print government node details
	fmt.Println("Government node details:")
	for _, record := range govNode {
		fmt.Printf("Node: %v\n", record["g"])
	}
}

File: neo4j_utils_main.py
#!/usr/bin/env python3

import sys
import os
import time

# Add the parent directory to sys.path
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from neo4j_util.neo4j_interface import Neo4jInterface

def create_government_node(neo4j_interface):
    """Create a government node with better error handling"""
    query = """
    MERGE (g:government {id: $id})
    SET g.name = $name
    RETURN g
    """
    params = {"id": "gov_01", "name": "Government of Sri Lanka"}
    
    max_attempts = 3
    for attempt in range(1, max_attempts + 1):
        try:
            print(f"Attempt {attempt} of {max_attempts} to create government node...")
            result = neo4j_interface.execute_query(query, params)
            print("Successfully created government node.")
            return result
        except Exception as e:
            print(f"Error on attempt {attempt}: {e}")
            if attempt < max_attempts:
                wait_time = 2 * attempt  # Increasing backoff
                print(f"Waiting {wait_time} seconds before retrying...")
                time.sleep(wait_time)
            else:
                print("Maximum retry attempts reached. Failed to create government node.")
                raise

def main():
    # Initialize Neo4j interface
    try:
        neo4j_interface = Neo4jInterface()
        
        # Create the government node
        gov_node = create_government_node(neo4j_interface)
        if gov_node:
            print("Government node details:")
            for record in gov_node:
                print(f"Node: {record['g']}")
                
    except Exception as e:
        print(f"Error in main function: {e}")
    finally:
        # Ensure we always close the connection
        try:
            neo4j_interface.close()
            print("Neo4j connection closed.")
        except Exception as e:
            print(f"Error closing Neo4j connection: {e}")

if __name__ == "__main__":
    main() 